app/src/main/java/pl/pwr/zpi/bcycle/mobile/RegisterActivity.kt

```diff
@@ -30,52 +36,178 @@ class RegisterActivity : AppCompatActivity() {
+                .start(this);
+
+        }
+
+        if(requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE) {
+
+            val result : CropImage.ActivityResult = CropImage.getActivityResult(data)
+
+            if(resultCode == Activity.RESULT_OK) {
+
+                mImageUri = result.uri
+
+                Picasso.get().load(mImageUri).into(avatarImV)
+
+            }
+
+        }
+    }
 
-    // TODO: Implement actual form validation
     private fun isFormValid(): Boolean =
-        namePT.content().isNotEmpty()
+     allDataFilled() && isPasswordOk() && isPasswordIdentical() && accountNotExists()
+
+    private fun allDataFilled(): Boolean {
+        if(namePT.content().isEmpty()) {
+            namePT.error = "empty!"
+        }
+        if(emailPT.content().isEmpty()) {
+            emailPT.error = "empty!"
+        }
+        if(passwordPT.content().isEmpty()) {
+            passwordPT.error = "empty!"
+        }
+        if(repeatPasswordPT.content().isEmpty()){
+            repeatPasswordPT.error = "empty!"
+        }
+
+        return namePT.content().isNotEmpty()
                 && emailPT.content().isNotEmpty()
                 && passwordPT.content().isNotEmpty()
-                && passwordPT.content() == repeatPasswordPT.content()
+                && repeatPasswordPT.content().isNotEmpty()
+    }
 
+    private fun accountNotExists() :Boolean {
+        return !auth.isSignInWithEmailLink(emailPT.content())
+    }
+
+    private fun isPasswordOk() : Boolean {
+        if(passwordPT.content().length <6 ){
+            passwordPT.error = "Password should be at least 6 characters!"
+        }
+        return passwordPT.content().length >5
+    }
+
+    private fun isPasswordIdentical() : Boolean {
+        if(passwordPT.content()!= repeatPasswordPT.content()) {
+            repeatPasswordPT.error ="different password"
+        }
+
+        return passwordPT.content()== repeatPasswordPT.content()
+    }
     private fun register() = auth
         .createUserWithEmailAndPassword(emailPT.content(), passwordPT.content())
         .addOnSuccessListener { updateUserDetails(it.user!!) }
-        .addOnFailureListener { showToast("Failed to register: ${it.localizedMessage}") }
+        .addOnFailureListener {
+
+            if (it.message.equals("The email address is already in use by another account.")) {
+                emailPT.error = "The email address is already in use by another account!"
+            }
+
+            if (it.message.equals("The email address is badly formatted.")) {
+                emailPT.error = "The email address is badly formatted!"
+            }
+        }
 
     private fun updateUserDetails(user: FirebaseUser) {
         val avatar: Uri? = if(uploadImageToFirebase()) {
-            Uri.parse("${FirebaseStorage.getInstance().reference.root}/${namePT.content()}.png")
+            Uri.parse("${FirebaseStorage.getInstance().reference.root}/${emailPT.content()}.png")
```

Comment:
If we use the user ID, we can allow users to change their e-mail address, and we don’t have to give out user e-mails to everyone when we display avatars in a trip. You can access the user ID with `user.uid`.